packages/loopring_v3/contracts/aux/access/LoopringIOExchangeOwner.sol

```diff
@@ -142,32 +147,49 @@ contract LoopringIOExchangeOwner is SelectorBasedAccessManager, ERC1271, Drainab
             require(blockIdx < blocks.length, "INVALID_BLOCKIDX");
             ExchangeData.Block memory _block = blocks[blockIdx];
 
-            _processTxCallbacks(_block, blockCallback.txCallbacks, callbackConfig.receivers);
+            for (uint j = 0; j < ctxs.length; j++) {
+               config.receivers[j].beforeEachBlock(_block, ctxs[j]);
+            }
+
+            _processTransactions(_block, ctxs, blockCallback.txs, config.receivers);
+
+            // for (uint j = 0; j < ctxs.length; j++) {
+            //     config.receivers[j].afterEachBlock(_block, ctxs[j]);
+            // }
+        }
+
+        for (uint i = 0; i < config.receivers.length; i++) {
+            config.receivers[i].afterAllBlocks(ctxs[i]);
         }
     }
 
-    function _processTxCallbacks(
-        ExchangeData.Block memory _block,
-        TxCallback[]       calldata txCallbacks,
-        address[]          calldata receivers
+    // Question(brecht): is `calldata` used correctly here?
```

Comment:
The big problem is the block here. Even in the most efficient scenario (which we're not in here, we still do costly memory->memory copies here generated by solidity) it's still expensive to pass through the complete block for each update (for a block with 512 transactions each call including the block will cost `(3+3)*(68/32)*512 = 6,528` gas just for the block data). And it's also pretty useless, we only use a couple of transactions, copying all the other transactions is useless work and costs gas.

To have lowest cost here we can just extract the transactions that are needed in `onAmmTransaction`. We can make a dummy `Block` object if we want just containing those couple of transactions (this way nothing needs to be changed in the pool contract) or just pass in the raw transaction data (that's the only thing we actually use currently I think). One extra input that we need to be able so that we can do that is the number of transactions that are expected to be processed in the callback to `TxCallback` (instead of just calculating that on the fly like we do now).